### DISCUSSION TITLE
VS code productivity and launch.json

### DISCUSSION THREAD

**Discussion by ChristianHinge** [ashleve/lightning-hydra-template#537] [Show and tell]
I am curious to hear your suggestions on vs-code productivity tips. I have customized my `launch.json` such that I can quickly select either `train` or `eval` and then input any overrides. Using the "Memento Inputs" extension, the arguments are saved and suggested when rerunning the specific debug mode.

![demo](https://user-images.githubusercontent.com/17361630/221126180-72fa4683-8797-4189-a47c-7049927eff88.gif)

My `launch.json`:
```
{
    "version": "0.2.0",
    "configurations": [
        {
            "name": "Python: Current File",
            "type": "python",
            "request": "launch",
            "program": "${file}",
            "console": "integratedTerminal",
            "justMyCode": true
        },
        {
            "name": "Python: Train X",
            "type": "python",
            "request": "launch",
            "program": "${workspaceFolder}/src/train.py",
            "args": "${input:trainArgs}",
            "console": "integratedTerminal",
            "justMyCode": true
        },
        {
            "name": "Python: Eval X",
            "type": "python",
            "request": "launch",
            "program": "${workspaceFolder}/src/eval.py",
            "args": "${input:evalArgs}",
            "console": "integratedTerminal",
            "justMyCode": true
        }
    ],
    "inputs": [
        {
        "id": "trainArgs",
        "type": "command",
        "command": "memento.promptString",
        "args":{
            "id": "trainArgs",
            "description": "Training",
            "default": "debug=fdr experiment="
            },
        },
        {
            "id": "evalArgs",
            "type": "command",
            "command": "memento.promptString",
            "args":{
                "id": "evalArgs",
                "description": "Eval",
                "default": "experiment="
                },
            }
    ]
}
```

**Comment by ashleve** [comment]
@ChristianHinge Thank you for tip! Very convenient indeed. I think I will start using this approach.

I'm not very deep into vs code, but I usually use `black` autoformatting on file save which saves me a lot of time during coding (it can be enabled somewhere in the vs code settings).

I wonder if there is any kind of minimal config, that could be added to template, which would allow for customizing plugins (like memento inputs) and general editor settings (like identation). Best if it could work with both vscode and pycharm. 

By "minimal" I mean something that won't override all your default configuration, but only the chosen part.

I've seen this editorconfig package which is a general config that works for many IDEs:
https://editorconfig.org/

But looking at the documentation, I don't see any mention about customizing specific plugins from different IDEs.

**Reply by ChristianHinge** [reply]
That could be quite cool. I haven't heard of something that works well across different editors. It might also cause trouble when editor extensions might not exist for all editors. For cookiecutter-based templates there could be the option of selecting an editor during project creation. However IMO, I like the simplicity of just cloning the template from git. If you feel like it would be valuable to add to the repo, perhaps it could be added to "tips and tricks" section for specific editors? Or a branch with editor-specific files akin to the wandb branch? :)
